Add a mechanism to GDScript to wait for signal's completion

This is feature request, and follows from the discussion in the issue #3223.

In GDScript, currently, there is no way to wait/block until a signal completes (in a general setting). This is in particular desirable in situations where the user wants to keep the game logic is synchronous, and the code structure flat (rather than sorting to a callback/futures/promises style game logic, which can be harder to develop and maintain).

It can be realized by adding a blocking alternative of signal-yield which blocks/waits until the signal is handled completely, or adding a `blocking=false` parameter to signal variant of yield for example.

@RandomShaper What I mean is something like this:
``` GDScript

func _process(delta):
    # some code
    do_something() # notice there is no yield
    # resumed synchronously

func do_something():
    # some code
    yield(do_something_2(), "completed")
    # resumed asynchronously when do_something_2 completes

func do_something_2():
    # some code
    yield(timer, "timeout")
    # resumed asynchronously on timeout
```

In C# the `async` keywords specifies that a method can use `await`. If the method does not use `await` when calling another async method the execution continues synchronously like in my example. A warning is printed by the compiler in such cases: _“warning CS4014: Because this call is not awaited, execution of the current method continues…”_

I get it now. Thank you for the example.

I'm very interested in seeing this supported (and helping to implement it
if needed), but also I think it's being wise to consider all the possible
design choices so this feature ends up being as good as possible.

El 23 abr. 2017 8:04 p. m., "Ignacio Etcheverry" <notifications@github.com>
escribió:

> @RandomShaper <https://github.com/RandomShaper> What I mean is something
> like this:
>
> func _process(delta):
>     # some code
>     do_something() # notice there is no yield
>     # resumed synchronously
> func do_something():
>     # some code
>     yield(do_something_2(), "completed")
>     # resumed asynchronously when do_something_2 completes
> func do_something_2():
>     # some code
>     yield(timer, "timeout")
>     # resumed asynchronously on timeout
>
> In C# the async keywords specifies that a method can use await. If the
> method does not use await when calling another async method the execution
> continues synchronously like in my example. A warning is printed by the
> compiler in such cases: *“warning CS4014: Because this call is not
> awaited, execution of the current method continues…”*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/issues/3235#issuecomment-296469040>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALQCtmhinwGCUbyBhAmrNPQgEjYCiWUAks5ry5KQgaJpZM4G-eFX>
> .
>

--- a/modules/gdscript/gd_function.cpp
+++ b/modules/gdscript/gd_function.cpp
@@ -1428,8 +1428,24 @@ Variant GDFunctionState::_signal_callback(const Variant **p_args, int p_argcount
 
 	state.result = arg;
 	Variant ret = function->call(NULL, NULL, 0, r_error, &state);
+
+	bool completed = true;
+
+	// If the return value is a GDFunctionState reference,
+	// then the function did yield again after resuming.
+	if (ret.is_ref()) {
+		GDFunctionState *gdfs = ret.operator Object *()->cast_to<GDFunctionState>();
+		if (gdfs && gdfs->function == function)
+			completed = false;
+	}
+
 	function = NULL; //cleaned up;
 	state.result = Variant();
+
+	if (completed) {
+		emit_signal("completed", ret);
+	}
+
 	return ret;
 }
 
@@ -1468,8 +1484,24 @@ Variant GDFunctionState::resume(const Variant &p_arg) {
 	state.result = p_arg;
 	Variant::CallError err;
 	Variant ret = function->call(NULL, NULL, 0, err, &state);
+
+	bool completed = true;
+
+	// If the return value is a GDFunctionState reference,
+	// then the function did yield again after resuming.
+	if (ret.is_ref()) {
+		GDFunctionState *gdfs = ret.operator Object *()->cast_to<GDFunctionState>();
+		if (gdfs && gdfs->function == function)
+			completed = false;
+	}
+
 	function = NULL; //cleaned up;
 	state.result = Variant();
+
+	if (completed) {
+		emit_signal("completed", ret);
+	}
+
 	return ret;
 }
 
@@ -1478,6 +1510,8 @@ void GDFunctionState::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("resume:Variant", "arg"), &GDFunctionState::resume, DEFVAL(Variant()));
 	ClassDB::bind_method(D_METHOD("is_valid", "extended_check"), &GDFunctionState::is_valid, DEFVAL(false));
 	ClassDB::bind_vararg_method(METHOD_FLAGS_DEFAULT, "_signal_callback", &GDFunctionState::_signal_callback, MethodInfo("_signal_callback"));
+
+	ADD_SIGNAL(MethodInfo("completed", PropertyInfo(Variant::NIL, "result")));
 }
 
 GDFunctionState::GDFunctionState() {

